Test Framework doc improvements (#34652)
Azure/azure-sdk-for-net: sdk/core/Azure.Core.TestFramework/README.md

</Project>
```

As an example, see the [Template](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/template/Azure.Template/tests/Azure.Template.Tests.csproj#L15) project. If you create a new project from the template, the Test Framework will be already referenced.

## Sync-async tests

> The Test Framework provides the ability to write tests using async client methods and automatically run them using sync overloads. This means that you don't need to duplicate tests to cover calling both the sync and async overloads of service client methods. To write sync-async client tests, inherit from `ClientTestBase` class and use the `InstrumentClient` method to wrap your client into a proxy class. This proxy class will automatically forward async calls to their sync overloads for the sync version of the test. `ClientTestBase` defines two TestFixtures - one for sync tests and one for async tests.

Author: Yeah I was going to talk more about test fixtures but then got distracted 😆 
I don't think it is necessary anyway.

In the test explorer, async tests will display as `TestClassName(true)` and sync tests as `TestClassName(false)`.

When using sync-async tests with recorded tests two sessions files will be generated - the async test session will have `Async.json` suffix. Note that if defining a recorded test, you would inherit from `RecordedTestBase` instead of `ClientTestBase`. This will be discussed in more detail in [Recorded Tests](#recorded-tests).

You can disable the sync-forwarding for an individual test by applying the `[AsyncOnly]` attribute to the test method.

__Limitation__: all method calls/properties that are being used have to be `virtual`.

## Recorded tests

> The bulk of the functionality of the Test Framework is around supporting the ability to run what we call recorded tests. This type of test can be thought of as a functional test as opposed to a unit test. A recorded test can be run in three different [modes](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.TestFramework/src/RecordedTestMode.cs):

Author: Not sure I agree with this - I wanted to distinguish between functional and unit tests. I think you are basically saying it kind of acts like a unit test that has its dependencies mocked when run in playback mode, but I'm not sure if that is helpful or confusing.

> The bulk of the functionality of the Test Framework is around supporting the ability to run what we call recorded tests. This type of test can be thought of as a functional test as opposed to a unit test. A recorded test can be run in three different [modes](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.TestFramework/src/RecordedTestMode.cs):

Reviewer: I don't feel strongly either way - but I like calling out that when run in playback mode, they function very similarly to unit tests.

> The bulk of the functionality of the Test Framework is around supporting the ability to run what we call recorded tests. This type of test can be thought of as a functional test as opposed to a unit test. A recorded test can be run in three different [modes](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.TestFramework/src/RecordedTestMode.cs):

Reviewer 2: I agree with Josh. Unit tests shouldn't incur all the cost of the recording framework and be in separate classes. Typically these are any parameter validation or error handling using mocked pipelines, which the recording framework will get in the way of anyway.

  - `Live` - The requests in the tests are run against live Azure resources.
  - `Record` - This is the same as live mode with one key difference - the HTTP traffic from your tests is saved locally on your machine.

>   - `Playback` - The requests that your library generates when running a test are compared against the requests in the recording for that test. For each matched request, the corresponding response is extracted from the recording and "played back" as the response.

Author: I will reword slightly to mention the caveats around matching/sanitization.

}
```

If more advanced sanitization is needed, you can use any of the regex-based sanitizer properties of `RecordedTestBase`. These are listed below along with example usages.
- [BodyKeySanitizers](https://grep.app/search?q=BodyKeySanitizers&filter[repo][0]=Azure/azure-sdk-for-net)
- [BodyRegexSanitizers](https://grep.app/search?q=BodyRegexSanitizers&filter[repo][0]=Azure/azure-sdk-for-net)
- [UriRegexSanitizers](https://grep.app/search?q=UriRegexSanitizers&filter[repo][0]=Azure/azure-sdk-for-net)
- [HeaderRegexSanitizers](https://grep.app/search?q=HeaderRegexSanitizers&filter[repo][0]=Azure/azure-sdk-for-net)

> _Note that when using any of the regex sanitizers, you must take care to ensure that the regex is specific enough to not match unintended values. When a regex is too broad and matches unintended values, this often manifests as the request or response being corrupted._

Author: Nothing specific but added more detail as to the exception that may result.

Once you have cloned the repo, open the [Test Proxy solution](https://github.com/Azure/azure-sdk-tools/blob/main/tools/test-proxy/Azure.Sdk.Tools.TestProxy.sln) in your IDE.

If you are attempting to debug `Playback` mode, set a breakpoint in the HandlePlaybackRequest method of [RecordingHandler](https://github.com/Azure/azure-sdk-tools/blob/main/tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs). If you are attempting to debug `Record` mode, set a breakpoint in the `HandleRecordRequestAsync` method of [RecordingHandler](https://github.com/Azure/azure-sdk-tools/blob/main/tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs). It may also be helpful to put breakpoints in [Admin.cs](https://github.com/Azure/azure-sdk-tools/blob/main/tools/test-proxy/Azure.Sdk.Tools.TestProxy/Admin.cs) to verify that your sanitizers are being added as expected.

With your breakpoints set, run the Test Proxy project, and then run your test that you are trying to debug. You should see your breakpoints hit. 

The key integration points between the Test Framework and the Test Proxy are:

>  - InstrumentClientOptions method of `RecordedTestBase` - calling this on your client options will set the [ClientOptions.Transport property](https://learn.microsoft.com/dotnet/api/azure.core.clientoptions.transport?view=azure-dotnet) to be [ProxyTransport](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.TestFramework/src/ProxyTransport.cs) to your client options. The ProxyTransport will send all requests to the Test Proxy.

Reviewer: ```suggestion
 - InstrumentClientOptions method of `RecordedTestBase` - calling this on your client options will set the [ClientOptions.Transport property](https://learn.microsoft.com/dotnet/api/azure.core.clientoptions.transport?view=azure-dotnet) to be [ProxyTransport](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.TestFramework/src/ProxyTransport.cs) to your client options when in `Playback` or `Record` mode. The ProxyTransport will send all requests to the Test Proxy.
```

If this is neglected, _clientDiagnostics will be null at test runtime.

## Miscellaneous

You can use `Recording.GenerateId()` to generate repeatable random IDs.

You should only use `Recording.Random` for random values (and you MUST make the same number of random calls in the same order every test run)

You can use `Recording.Now` and `Recording.UtcNow` if you need certain values to capture the time the test was recorded.

It's possible to add additional recording variables for advanced scenarios (like custom test configuration, etc.) by using `Recording.SetVariable` or `Recording.GetVariable`.

You can use `if (Mode == RecordingMode.Playback) { ... }` to change behavior for playback only scenarios (in particular to make polling times instantaneous)

You can use `using (Recording.DisableRecording()) { ... }` to disable recording in the code block (useful for polling methods)

> In order to enable testing with Fiddler, you can either set the  `AZURE_ENABLE_FIDDLER` environment variable or the `EnableFiddler` [runsetting](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/nunit.runsettings) parameter to `true`.

Reviewer 2: Nit: feels more like a bullet list here would read better.

If this is neglected, _clientDiagnostics will be null at test runtime.

## Miscellaneous

You can use `Recording.GenerateId()` to generate repeatable random IDs.

You should only use `Recording.Random` for random values (and you MUST make the same number of random calls in the same order every test run)

> You can use `Recording.Now` and `Recording.UtcNow` if you need certain values to capture the time the test was recorded.

Reviewer: ```suggestion
You can use `Recording.Now` and `Recording.UtcNow` if you need to use date or time values that will be included in the recording.
```

It's possible to add additional recording variables for advanced scenarios (like custom test configuration, etc.) by using `Recording.SetVariable` or `Recording.GetVariable`.

You can use `if (Mode == RecordingMode.Playback) { ... }` to change behavior for playback only scenarios (in particular to make polling times instantaneous)

You can use `using (Recording.DisableRecording()) { ... }` to disable recording in the code block (useful for polling methods)

> In order to enable testing with Fiddler, you can either set the  `AZURE_ENABLE_FIDDLER` environment variable or the `EnableFiddler` [runsetting](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/nunit.runsettings) parameter to `true`.

Reviewer: ```suggestion
In order to observe test network traffic with Fiddler, you can either set the `AZURE_ENABLE_FIDDLER` environment variable or the `EnableFiddler` [runsetting](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/nunit.runsettings) parameter to `true`.
```
